### PR TITLE
Add support for `*-deprecation` command-line flags of Node.js

### LINKED ISSUE #7533: Use `process.emitWarning()` instead of `console.warn()`
### What is the problem you're trying to solve?

Currently, Stylelint uses [`console.warn()`](https://nodejs.org/api/console.html#consolewarndata-args) to show deprecation warnings, e.g.

https://github.com/stylelint/stylelint/blob/c49f9b5a7b45cb35c58b5bc1cd669733896d0183/lib/augmentConfig.mjs#L325-L327

`console.warn()` is an alias for `console.error()` in Node.js and just prints a message to `stderr`. There is no easy way to handle the output, e.g. stop printing.

### What solution would you like to see?

How about using [`process.emitWarning()`](https://nodejs.org/api/process.html#processemitwarningwarning-options) instead? E.g.

```js
emitWarning( 
	`CommonJS plugins are deprecated ("${pluginLookup}"). See https://stylelint.io/migration-guide/to-16`,
	{ type: 'DeprecationWarning', code: 'stylelint:001' }
);
```

This function of `node:process` provides a common way to handle deprecation warnings in a Node.js program. `DeprecationWarning` is a reserved type and affects the `--*-deprecation` command-line flags of Node.js like `--no-deprecation`. For details, see <https://nodejs.org/api/process.html#event-warning>.

E.g.

```sh-session
$ npx stylelint '*.css'
(node:20670) [stylelint:001] DeprecationWarning: CommonJS plugins are deprecated ("foo"). See https://stylelint.io/migration-guide/to-16
(Use `node --trace-deprecation ...` to show where the warning was created)
```

Using the `--no-deprecation` flag can suppress the warning:

```sh-session
$ NODE_OPTIONS='--no-deprecation' npx stylelint '*.css'
```

Please let me know if anyone knows some downsides of `process.emitWarning()`.

### PR DESCRIPTION
**Pull request by fpetrakov**
> Which issue, if any, is this issue related to?

Closes #7533

> Is there anything in the PR that needs further explanation?

No, it's self-explanatory.

### CODE REVIEW

File: lib/augmentConfig.mjs [hunk 1]
```diff
@@ -3,6 +3,7 @@ import { dirname, isAbsolute } from 'node:path';
 import globjoin from 'globjoin';
 import micromatch from 'micromatch';
 import normalizePath from 'normalize-path';
+import process from 'process';
```
Comment (Mouvedia) [inline lib/augmentConfig.mjs:6]: nit

`'node:process'` for consistency

File: lib/augmentConfig.mjs [hunk 2]
```diff
@@ -322,7 +323,7 @@ async function addPluginFunctions(config, { quietDeprecationWarnings }) {
 
 			// NOTE: This '.cjs' check is limited. Some CommonJS plugins may have the '.js' extension.
 			if (!quietDeprecationWarnings && pluginLookup.endsWith('.cjs')) {
-				console.warn(
+				process.emitWarning(
 					`CommonJS plugins are deprecated ("${pluginLookup}"). See https://stylelint.io/migration-guide/to-16`,
 				);
```
Comment (Mouvedia) [inline lib/augmentConfig.mjs:328]:
```suggestion
				process.emitWarning(`CommonJS plugins are deprecated ("${pluginLookup}")`, {
				    type: 'DeprecationWarning',
				    code: 'stylelint:001',
				    detail: 'See https://stylelint.io/migration-guide/to-16'
				});
```

File: rollup.config.mjs [hunk 3]
```diff
@@ -64,7 +64,7 @@ function addWarningForCommonJS() {
 					'	// [INSERT HERE] CommonJS deprecation code',
 					`
 	if (!options.quietDeprecationWarnings) {
-		console.warn('The CommonJS Node.js API is deprecated. See https://stylelint.io/migration-guide/to-16');
+		process.emitWarning('The CommonJS Node.js API is deprecated. See https://stylelint.io/migration-guide/to-16');
```
Comment (Mouvedia) [inline rollup.config.mjs:67]:
```suggestion
		process.emitWarning(`CommonJS plugins are deprecated ("${pluginLookup}")`, {
		  type: 'DeprecationWarning',
		  code: 'stylelint:001',
		  detail: 'See https://stylelint.io/migration-guide/to-16'
		});
```

Comment (Mouvedia) [inline rollup.config.mjs:67]:
```suggestion
		process.emitWarning('The CommonJS Node.js API is deprecated', {
		  type: 'DeprecationWarning',
    		  code: 'stylelint:002',
    		  detail: 'See https://stylelint.io/migration-guide/to-16'
		});
```

File: lib/augmentConfig.mjs (hunk 2 above)
Comment (Mouvedia) [inline lib/augmentConfig.mjs:328]:
```suggestion
				process.emitWarning(`CommonJS plugins are deprecated ("${pluginLookup}")`, {
				  type: 'DeprecationWarning',
				  code: 'stylelint:001',
				  detail: 'See https://stylelint.io/migration-guide/to-16'
				});
```

File: lib/prepareReturnValue.mjs [hunk 4]
```diff
@@ -35,7 +37,7 @@ export default function prepareReturnValue(stylelintResults, maxWarnings, format
 		// TODO: Deprecated. Remove in the next major version.
 		get output() {
 			if (!this._outputWarned) {
-				console.warn('`output` is deprecated. Use `report` or `code` instead.');
+				process.emitWarning('`output` is deprecated. Use `report` or `code` instead.');
```
Comment (Mouvedia) [inline lib/prepareReturnValue.mjs:40]:
```suggestion
				process.emitWarning('`output` is deprecated', {
				  type: 'DeprecationWarning',
				  code: 'stylelint:003',
				  detail: 'Use `report` or `code` instead.'
				});
```

File: lib/__tests__/standalone-deprecations.test.mjs [hunk 5]
```diff
@@ -1,3 +1,5 @@
+import process from 'process';
```
Comment (ybiquitous) [inline lib/__tests__/standalone-deprecations.test.mjs:1]:
```suggestion
import process from 'node:process';
```

File: lib/augmentConfig.mjs (hunk 1 above)
Comment (ybiquitous) [reply lib/augmentConfig.mjs:6]: Agree. I want to distinguish the standard library and 3rd-party libraries clearly in imports.

**Review by ybiquitous — CHANGES_REQUESTED** [review]
@fpetrakov Thanks for the pull request!

As @Mouvedia commented, `'DeprecationWarning'` is required to affect the `*-deprecation` CLI flags of Node.js. Can you address the comment, please?

Ref https://nodejs.org/api/process.html#event-warning

**Comment by ybiquitous** [comment]
@Mouvedia Thanks for the good points 👍🏼 

I think having a period is better as the Node.js API does:

```sh-session
$ node -e "new Buffer('')" --throw-deprecation
node:internal/process/warning:185
        throw warning;
        ^

DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
    at showFlaggedDeprecation (node:buffer:188:11)
    at new Buffer (node:buffer:266:3)
    at [eval]:1:1
    at runScriptInThisContext (node:internal/vm:209:10)
    at node:internal/process/execution:109:14
    at [eval]-wrapper:6:24
    at runScript (node:internal/process/execution:92:62)
    at evalScript (node:internal/process/execution:123:10)
    at node:internal/main/eval_string:51:3 {
  code: 'DEP0005'
}

Node.js v21.7.1
```

Also, I'm fine if the code order is like `stylelint:001`, `stylelint:002`, ... 👍🏼 

Ref https://nodejs.org/api/deprecations.html

---

Besides, although it is not required, adding a utility might be helpful, e.g.

```js
const CODES = {
  '001': {message: 'Message 1', detail: '...'},
  '002': {message: 'Message 2'},
  // ...
};

/**
 * @param {keyof CODES} code
 */
function emitDeprecation(code) {
  const {message, detail} = CODES[code];
  process.emitWarning(message, {type: 'DeprecationWarning', code, detail});
}
```

**Comment by Mouvedia** [comment]
> I think having a period is better as…

The suggestions have been updated.

> I'm fine if the code order…

OK

**Comment by Mouvedia** [comment]
@fpetrakov @ybiquitous `npm build` needs to be run again after the code review suggestions commit.

**Comment by ybiquitous** [comment]
@Mouvedia Do you have any thoughts left?

**Review by Mouvedia — APPROVED** [review]
> @Mouvedia Do you have any thoughts left?

Maybe add a task to #7396 that says we need to run tests using `--throw-deprecation` before releasing. For the rest it LGTM.

**Comment by ybiquitous** [comment]
> Maybe add a task to #7396 that says we need to run tests using `--throw-deprecation` before releasing.

Sounds good. Can you open an issue and add it to the list?

**Pull request merged**